src/document.ts

```diff
@@ -1,17 +1,21 @@
 import rfdc from 'rfdc';
+import type Model from './model';
+import type Schema from './schema';
+import type { NodeJSLikeCallback } from './types';
 const cloneDeep = rfdc();
 
-abstract class Document {
-  abstract _model;
-  _id!: any;
-  abstract _schema;
+abstract class Document<T> {
+  abstract _model: Model<T>;
+  _id!: string | number | undefined;
+  abstract _schema: Schema;
+  [key : string]: any;
```

Comment:
I'm not sure how this can be modified to use the indexes in the generalized type.

Reply:
maybe https://stackoverflow.com/questions/69291358/shortcut-syntax-to-class-constructor-from-typescript-interface/69293235#69293235 can help